### PR TITLE
Avoid cache clearing when user claims are not updated.

### PR DESCRIPTION
**Pull request by ravinduramesh**
### Proposed changes in this pull request
$Subject

### Related Issues
- https://github.com/wso2/product-is/issues/15127

### Related PRs
Merge with the following PRs:
- https://github.com/wso2/carbon-identity-framework/pull/4344

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 1]
```diff
@@ -100,6 +103,16 @@ public boolean doPreUpdateApplication(ServiceProvider serviceProvider, String te
         handleOAuthAppAssociationRemoval(serviceProvider);
         storeSaaSPropertyValue(serviceProvider);
         removeClientSecret(serviceProvider);
+
+        ServiceProvider storedSP = OAuth2ServiceComponentHolder.getApplicationMgtService()
+                .getServiceProvider(serviceProvider.getApplicationID());
+        ClaimConfig updatedClaimConfig = serviceProvider.getClaimConfig();
+        updatedClaimConfig.setSpClaimDialects(new String[0]);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:110]: Won't this override any existing value if set?

Comment (mefarazath) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:110]: Let's say for SAML SPs?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 2]
```diff
@@ -100,6 +102,14 @@ public boolean doPreUpdateApplication(ServiceProvider serviceProvider, String te
         handleOAuthAppAssociationRemoval(serviceProvider);
         storeSaaSPropertyValue(serviceProvider);
         removeClientSecret(serviceProvider);
+
+        ServiceProvider storedSP = OAuth2ServiceComponentHolder.getApplicationMgtService()
+                .getServiceProvider(serviceProvider.getApplicationID());
+        String updatedClaimConfigString = new Gson().toJson(serviceProvider.getClaimConfig());
+        String storedClaimConfigString = new Gson().toJson(storedSP.getClaimConfig());
+        if (updatedClaimConfigString.equals(storedClaimConfigString)) {
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:110]: User STring.utils to compare

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 3]
```diff
@@ -100,6 +102,14 @@ public boolean doPreUpdateApplication(ServiceProvider serviceProvider, String te
         handleOAuthAppAssociationRemoval(serviceProvider);
         storeSaaSPropertyValue(serviceProvider);
         removeClientSecret(serviceProvider);
+
+        ServiceProvider storedSP = OAuth2ServiceComponentHolder.getApplicationMgtService()
+                .getServiceProvider(serviceProvider.getApplicationID());
+        String updatedClaimConfigString = new Gson().toJson(serviceProvider.getClaimConfig());
+        String storedClaimConfigString = new Gson().toJson(storedSP.getClaimConfig());
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:109]: change stored -> existing

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 4]
```diff
@@ -100,6 +102,14 @@ public boolean doPreUpdateApplication(ServiceProvider serviceProvider, String te
         handleOAuthAppAssociationRemoval(serviceProvider);
         storeSaaSPropertyValue(serviceProvider);
         removeClientSecret(serviceProvider);
+
+        ServiceProvider storedSP = OAuth2ServiceComponentHolder.getApplicationMgtService()
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:106]: change stored -> existing

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 5]
```diff
@@ -132,7 +142,11 @@ public boolean doPostUpdateApplication(ServiceProvider serviceProvider, String t
         revokeAccessTokensWhenSaaSDisabled(serviceProvider, tenantDomain);
         addClientSecret(serviceProvider);
         updateAuthApplication(serviceProvider);
-        removeEntriesFromCache(serviceProvider, tenantDomain);
+
+        if (threadLocalForClaimConfigUpdates.get()) {
+            removeEntriesFromCache(serviceProvider, tenantDomain);
+        }
+        threadLocalForClaimConfigUpdates.remove();
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:149]: Clear this in finally block

**Review by inthirakumaaran — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 6]
```diff
@@ -100,6 +102,14 @@ public boolean doPreUpdateApplication(ServiceProvider serviceProvider, String te
...
         storeSaaSPropertyValue(serviceProvider);
         removeClientSecret(serviceProvider);
+
+        ServiceProvider existingSP = OAuth2ServiceComponentHolder.getApplicationMgtService()
+                .getServiceProvider(serviceProvider.getApplicationID());
+        String newClaimConfigString = new Gson().toJson(serviceProvider.getClaimConfig());
+        String existingClaimConfigString = new Gson().toJson(existingSP.getClaimConfig());
+        if (StringUtils.equals(newClaimConfigString, existingClaimConfigString)) {
+            threadLocalForClaimConfigUpdates.set(false);
+        }
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:112]: add a else block and make the Thread local to true.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java [hunk 7]
```diff
@@ -100,6 +102,14 @@ public boolean doPreUpdateApplication(ServiceProvider serviceProvider, String te
         handleOAuthAppAssociationRemoval(serviceProvider);
         storeSaaSPropertyValue(serviceProvider);
         removeClientSecret(serviceProvider);
+
+        ServiceProvider existingSP = OAuth2ServiceComponentHolder.getApplicationMgtService()
+                .getServiceProvider(serviceProvider.getApplicationID());
+        String newClaimConfigString = new Gson().toJson(serviceProvider.getClaimConfig());
```
Comment (ravinduramesh) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuthApplicationMgtListener.java:108]: If the order get changed, string comparison will be failed. Therefore, compare two objects.

**Pull request merged**